torch/sparse/semi_structured.py

```diff
@@ -0,0 +1,362 @@
+    "_SEMI_STRUCTURED_SPARSE_CONFIG", "compression_factor n min_size"
+)
+_DTYPE_TO_SEMI_STRUCTURED_SPARSE_CONFIG = {
+    torch.float16: _SEMI_STRUCTURED_SPARSE_CONFIG(9, 2, 64),
+    torch.int8: _SEMI_STRUCTURED_SPARSE_CONFIG(10, 2, 128),
+}
+
+
+class SparseSemiStructuredTensor(torch.Tensor):
+    """This class implementes semi-structured sparsity as a Tensor subclass.
+
+    Semi-structured sparsity describes a sparsity pattern where n in every 2n elements are sparse,
+    depending on the datatype. It is also referred to as 2:4 sparsity or fine-grained
+    structured sparsity.
+
+    Currently, this class supports 2:4 sparsity for int8 and float16 dtypes.
+
+    This subclass stores the dense tensor in a compressed form by only storing the specified elemenets and a metadata mask.
+    These two are stored next to each other in one contiguous tensor.
+
+    We choose to store the specified elements and the metadata in a single tensor for future compatibilty with cuSPARSELt.
+
+    compressed tensor = [ specified elements of original tensor |   mask_metadata     ]
+
+    For an original tensor of size (m, k) we expect the first m * k // 2 elements to be the kept elements
+    The rest of the tensor is metadata.
+
+    This subclass also overrides __torch_dispatch__ to use _structured_sparse_linear for faster matrix multiplications
+    via sparse CUTLASS kernels. In the future we will also call into cuSPARSELt kernels for more performance gains.
+    """
+
+    @staticmethod
+    def __new__(cls, custom_shape: torch.Size, compressed_tensor: torch.Tensor, transposed: bool = False):
+        """
+        Create a new instance of the class.
+
+        Args:
+            custom_shape (tuple): The custom shape for the new instance.
+            compressed_tensor (torch.Tensor): The compressed tensor to use for the new instance.
+            transposed (bool): Indicates whether the tensor is transposed.
+
+        Returns:
+            torch.Tensor: A torch.Tensor wrapper subclass.
+
+        Raises:
+            None
+
+        """
+        kwargs = {}
+        kwargs["device"] = compressed_tensor.device  # type: ignore[assignment]
+        kwargs["dtype"] = compressed_tensor.dtype  # type: ignore[assignment]
+        kwargs["layout"] = compressed_tensor.layout  # type: ignore[assignment]
+        kwargs["requires_grad"] = False  # type: ignore[assignment]
+
+        return torch.Tensor._make_wrapper_subclass(cls, custom_shape, **kwargs)  # type: ignore[attr-defined]
+
+    def __init__(
+        self,
+        custom_shape: torch.Size,
+        compressed_tensor: torch.Tensor,
+        transposed: bool,
+    ) -> None:
+        """SparseSemiStructuredTensor constructor.
+
+        Args:
+            custom_shape: The shape of the original dense tensor
+            compressed_tensor: A flattened tensor to store the specified elements and mask metadata.
+            transposed: Whether the tensor is transposed or not.
```

Comment:
This only makes sense for 2D Tensors no?

Reply:
Yes, currently we are limited to only supporting 2D tensors

Reply:
Can you add an assert in this new or in the init to make sure this is enforced?

Reply:
I have all of the asserts for checking dtype, dim=2d, device in `to_sparse_semi_structured`, should I move all of those into the constructor?

Reply:
Moving the logic to the class to avoid duplication sounds good!